payment_gateway_adyen/services/payment_service.py

```diff
@@ -0,0 +1,173 @@
+        transaction = None
+        try:
+            result = self._get_adyen_client().authorise3d(request=payload)
+            vals = {
+                'state': MAP_SOURCE_STATE[result.message['resultCode']],
+                'data': json.dumps(result.message),
+            }
+
+            transaction = self.env['gateway.transaction'].search([
+                ('external_id', '=', result.message['pspReference']),
+                ('payment_mode_id.provider', '=', 'adyen'),
+                ])
+        except ValueError as e:
+            # NOTE catch orther errors?
+            # do we want to cancel the transaction? Can we do it?
+            raise UserError(e.message)
+
+        if transaction:
+            transaction.write(vals)
+        else:
+            raise UserError(
+                _('The transaction %s do not exist in Odoo') % result.message[
+                    'pspReference'])
+
+    def _get_formatted_amount(self, amount=None):
+        if amount is None:
+            amount = self.collection._get_amount_to_capture()
+        if self.collection.currency_id.name in ZERO_DECIMAL_CURRENCIES:
+            return int(amount)
+        elif self.collection.currency_id.name in THREE_DECIMAL_CURRENCIES:
+            return int(float_round(amount * 1000, 0))
+        else:
+            return int(float_round(amount * 100, 0))
+
+    def _prepare_charge(self, source=None, **kwargs):
+        transaction = self.collection
+        description = "|".join([
+            transaction.name,
+            transaction.partner_id.email,
+            str(transaction.id)])
+        vals = {
+            'amount': {
+                "value": self._get_formatted_amount(),
+                "currency": transaction.currency_id.name},
+            'reference': description,
+            'additionalData': {"executeThreeD": "true"}  # TODO optional?
+            }
+        if kwargs.get('card'):
+            # PCI clear card data
+            vals['card'] = kwargs['card']
+        elif kwargs.get('encrypted_card'):
+            vals['additionalData'] = {
+                'card.encrypted.json': kwargs['encrypted_card'],
+                'executeThreeD': 'true'
+            }
+        return vals
+
+    def _get_adyen_client(self):
+        account = self._get_account()
+        data = account.get_data()
+        ady = Adyen.Adyen()
+        ady.payment.client.username = data['username']
+        ady.payment.client.platform = str(data['platform'])
+        ady.payment.client.merchant_account = data['merchant_account']
+        ady.payment.client.password = account._get_password()
+        ady.payment.client.app_name = data['app_name']
+        return ady.payment
+
+    def _create_transaction(self, source=None, **kwargs):
+        try:
+            return self._get_adyen_client().authorise(
+                request=self._prepare_charge(source=source, **kwargs))
+        except ValueError as e:  # TODO catch orther errors?
```

Comment:
Can you add a test when the identifier are wrong, and try to raise a better error for that case?

Reply:
yep